### PR TITLE
Functional tbmv

### PR DESCRIPTION
**Pull request by daineAMD**
Functional s/d/c/z versions of L2 BLAS function tbmv. Un-optimized. Implementation is based on gemv.
Tbmv is important for implementation of tbsv which is needed in the future.

Tbmv is a triangular banded matrix - vector multiply. See comments in code for explanation, or see LAPACK.

SWDEV-214265

### CODE REVIEW

File: library/src/blas2/rocblas_tbmv.hpp [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        {
+            // Regular case, simply multiply
+            if(row < k && row > 0)
+            {
+                res_A += (A[row + col * lda] * x_copy[col]);
+            }
+            else if(row == 0)
+            {
+                // If main diagonal && diag, don't reference matrix, assume 1.
+                if(diag && !upper)
```
Comment (TorreZuk) [inline library/src/blas2/rocblas_tbmv.hpp]: minor tweak, not sure if it worth it 
if (diag && (!upper || (k==0 && upper)))

File: library/src/blas2/rocblas_tbmv.hpp [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    if(thread_id < DIM_X)
+    {
+        // Add the partial sums of each diagonal and store
+        for(rocblas_int i = 1; i < DIM_Y; i++)
+        {
+            sdata[thread_id] += sdata[thread_id + DIM_X * i];
+        }
+
+        // Update x.
+        if(ind < m)
```
Comment (TorreZuk) [inline library/src/blas2/rocblas_tbmv.hpp]: Makes no difference if this check is && with thread_id < DIM_X

File: clients/include/testing_tbmv.hpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        {
+            rocblas_tbmv<T>(handle, uplo, transA, diag, M, K, dA, lda, dx, incx);
+        }
+
+        gpu_time_used     = (get_time_us() - gpu_time_used) / number_hot_calls;
+        rocblas_gflops    = tbmv_gflop_count<T>(M, K) / gpu_time_used * 1e6;
+        rocblas_bandwidth = (1.0 * M * M) * sizeof(T) / gpu_time_used / 1e3;
+
+        // only norm_check return an norm error, unit check won't return anything
+        std::cout << "M,K,lda,incx,rocblas-Gflops,rocblas-GB/s,";
```
Comment (mahmoodw) [inline clients/include/testing_tbmv.hpp]: Nice to include time as well for completeness

File: library/src/blas2/rocblas_tbmv.hpp [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+ * ************************************************************************ */
+#ifndef __ROCBLAS_TBMV_HPP__
+#define __ROCBLAS_TBMV_HPP__
+#include "handle.h"
+#include "rocblas.h"
+#include "utility.h"
+
+// Makes a copy of xa into ya.
+template <typename T, typename U, typename V>
+__global__ void tbmv_copy_helper(rocblas_int    n,
```
Comment (TorreZuk) [inline library/src/blas2/rocblas_tbmv.hpp]: Could consider moving this to a shared location for reuse.

File: library/include/rocblas-functions.h [hunk 5]
```diff
@@ -3026,6 +3026,123 @@ ROCBLAS_EXPORT rocblas_status rocblas_zgemv_strided_batched(rocblas_handle
...
+              if uplo == rocblas_fill_upper:
+                The matrix represented is an upper banded triangular matrix
+                with the main diagonal and k super-diagonals, everything
+                else can be assumed to be 0.
+                The matrix is compacted so that the main diagonal resides on the k'th
+                row, the first super diagonal resides on the RHS of the k-1'th row, etc,
+                with the k'th diagonal on the RHS of the 0'th row.
+                   Ex: (rocblas_fill_upper; m = 5; k = 2)
+                      1 6 9 0 0              0 0 9 8 7
+                      0 2 7 8 0              0 6 7 8 9
```
Comment (mahmoodw) [inline library/include/rocblas-functions.h]: should this be 6782?

File: clients/include/testing_tbmv.hpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        if(arg.norm_check)
+        {
+            rocblas_error_1 = norm_check_general<T>('F', 1, M, abs_incx, hx_gold, hx_1);
+        }
+    }
+
+    if(arg.timing)
+    {
+        int number_cold_calls = 2;
+        int number_hot_calls  = 100;
```
Comment (mahmoodw) [inline clients/include/testing_tbmv.hpp]: Should be set to arg.iters to give rocblas-bench control of iterations - should fix gemv too

File: library/src/blas2/rocblas_tbmv.hpp (hunk 1 above)
Comment (daineAMD) [reply library/src/blas2/rocblas_tbmv.hpp]: Yeah that's a bit cleaner. I'm guessing a lot of this logic can be improved in the future. Changed.

File: library/src/blas2/rocblas_tbmv.hpp (hunk 4 above)
Comment (daineAMD) [reply library/src/blas2/rocblas_tbmv.hpp]: This was a straight up copy of the copy_kernel from L1 blas. I changed it to include rocblas_copy.hpp and call that directly.

**Pull request merged**